### ISSUE TITLE
Idemptency when sending events (avoid duplicate events)

### ISSUE THREAD

**Issue by alekseyrozh** [PostHog/posthog#17211]
## Is your feature request related to a problem?

Duplicate events possible when sending to posthog. My exact usecase is - i'm listening to appstore subscription purchase/renew events on my server and I want send those events to posthog to keep track of sales and MRR. However it is possible to end up with sending the same event to posthog twice:

* My server recieves notification about subscription renewal
* My server send corresponding event to posthog
* My server crashes
* Appstore will attempt to send the same notification with the same "NotificationUUID"  again to my server as server didn't respond with 200 when processing this request
* My server will end up sending the same event to posthog for the second time, which is incorrect and messes up the subscriptions and sales data 

## Describe the solution you'd like

Some way to ensure idempotency when sending events to posthog so on the posthog side there's no duplicates, for example:

* Providing idempotency key
```
posthog.capture({
    idempotencyKey: 'MY_IDEMPOTENCY_KEY
    distinctId: 'distinct_id_of_the_user',
    event: 'user paid for subscription',
})
```

* Or generating event id on a sender side
```
posthog.capture({
    eventId: 'MY_GENERATED_EVENT_ID'
    distinctId: 'distinct_id_of_the_user',
    event: 'user paid for subscription"',
})
```

## Describe alternatives you've considered

You can persist those unique IDs in the db before sending to posthog, but then there's a danger of never sending an event to Posthog if the server crashes after you persist to db but before you send event

## Additional context

I'm sure this property will be useful not only in server-server events but also client-server.

**Comment by MrTin** [comment]
I am chiming in here for another perhaps more common use case. When a user refreshes the page (either manually or by closing and opening their browser at a later time) we don't want to record multiple 'purchase' events in our case.

**Comment by tiina303** [comment]
We deduplicate based on the sort key [timestamp, distinct_id, event, uuid] <- if these fields are the same, then there will be eventual deduplication, so the solution here is to make sure you send the events with the same uuid.

**Comment by helloworld9912** [comment]
very usefull feature, can be usefull when importing old data and we have already live data reporting on to avoid duplicated events.
maybe a "distinctEventId" key?

**Comment by alekseyrozh** [comment]
@tiina303 could you please provide more details or some docs perhaps? what is the schema you mentioned in your reply, is this how an event that is supposed to sent to posthog? maybe you have some exampled at hand

**Comment by dan-dr** [comment]
> We deduplicate based on the sort key [timestamp, distinct_id, event, uuid] <- if these fields are the same, then there will be eventual deduplication, so the solution here is to make sure you send the events with the same uuid.

@tiina303 just uuid or all the fields?

**Comment by tiina303** [comment]
all of the fields need to match (event is the event name, e.g. pageview).

**Comment by alekseyrozh** [comment]
@tiina303 that means just explicitely setting uuid=some_id_here and timestamp=123123 when sending an event?

**Comment by tiina303** [comment]
yes, and make sure the event and distinct_id aren't changing (which would be odd).

**Comment by Darren120** [comment]
need to have a option to force dedupe.

**Comment by SethThoburn** [comment]
@tiina303 Can you explain what "eventual deduplication" means? What happens if I generate an insight before that deduplication happens?

**Comment by tiina303** [comment]
This is something we get from ClickHouse and there are no concrete guarantees here afaik. In practice we've seen this work quite fast especially for events sent near the same time. If deduplication hasn't happened yet you'd see the count in insights include the dupe.

**Comment by jclusso** [comment]
@tiina303 you mention sending a UUID, but there is no concept of this in the [Ruby SDK](https://www.rubydoc.info/gems/posthog-ruby/2.3.0/PostHog/Client#capture-instance_method). Would message_id be what you're referring to?

**Comment by jclusso** [comment]
After adding support to the Ruby library above, I was able to pass UUID as long as it's formated as a proper UUID and it eventually deduplicated. I was wondering if you fire the same event 5 times with different properties, what properties would be on the deduplicated event?

**Comment by xvello** [comment]
@jclusso this would be an interesting edge case, as our deduplication mechanism is designed for idempotency on identical events. As a general rule, the last event to be processed will be kept after deduplication, but there are several caveats to keep in mind:
- deduplication only happens if many fields match (currently: the project, day, event name, distinct_id and uuid). If any of these are different, deduplication will not trigger on uuid alone. But different event properties would indeed allow deduplication to happen.
- while our ingestion pipeline currently processes the vast majority of events in the order they are submitted, this is not a behaviour we’ll give strong guarantees about: projects [overflowing capacity](https://posthog.com/docs/data#event-ingestion-has-overflowed-capacity) can be processed out of order for example. If your 5 events are submitted as part of a big burst of traffic, we will not specify which one will be processed last.

**Comment by xvello** [comment]
@jclusso this deduplication mechanism is designed to mitigate cases where the client submits the same event several times, either:

- due to unreliable connections (it is transparently implemented in our web and mobile SDKs)
- due to an event source that does not guarantee idempotency (webhook, kafka), and only if a valid [UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier) is available: not all 128-bit words are valid UUIDs.

To reply to your question about submitting the same UUID with different properties: we will not guarantee or document the behavior of our systems if that happens, as this use case is outside of our design parameters. Assuming that the timestamp, event name and UUID match, a single event will be kept after deduplication, but we won't guarantee which one will be chosen.

**Comment by jclusso** [comment]
Thanks for that @xvello. I think it would be ideal if you guys improve your public documentation on this rather than people having to dig into various issues to find out how deduplication works. There are many times outside of unreliable connections that can cause a system to trigger a duplicate event and honestly it makes sense to always include a UUID from the client side if you at all care about idempotency.

**Comment by xvello** [comment]
Agreed @jclusso, we are currently discussing documentation changes to clarify the supported use cases, aligned with my latest comment. We also plan to provide automatic UUID generation on all SDKs that transparently queue and retry events, but cannot commit on a timeframe for this.

**Comment by klunejko** [comment]
@xvello How would one specify the uuid using posthog-js (web)? And having a field event_id or similar, that's used to automatically deduplicate events would be incredibly useful in either case. Otherwise we would have to create uuids solely so we can use those to send with each event. Something like Meta's conversion API would be ideal and simple to implement.

**Comment by Adonis0123** [comment]
After testing, specifying timestamp can solve the problem.

`posthog.capture('xxx', params, { timestamp })`

**Comment by hel-lo7** [comment]
I've provided the same uuid for duplicate events, but the events are not deduplicated and still appear in all analytics. I see in my data management tab that the uuid of the two events are the exact same. Is there anything else I need to do to enable deduplication?

**Comment by MarconLP** [comment]
Hey @hel-lo7, it might take a week or two for the de-duplication to kick in. We only de-duplicate events during non-peak hours such as weekends.

**Comment by kcrandall** [comment]
How is a uuid sent in the REST API? Is it put in the properties dictionary? Or next to the distinct_id? I couldn't find it specified in the /capture/ docs

**Comment by MarconLP** [comment]
Hey @kcrandall, the UUID has to be next to the distinct_id. Here is an example:
```
curl -v -L --header "Content-Type: application/json" -d '{
    "api_key": "<ph_project_api_key>",
    "distinct_id": "distinct_id_of_the_user",
    "event": "user_signed_up",
    "uuid": "01916987-08f2-0000-7f29-91ed6eeeeeee"
}' https://us.i.posthog.com/capture/
```

**Comment by marandaneto** [comment]
https://github.com/PostHog/posthog-js/pull/3547
